brainiak/utils/fmrisim.py

```diff
@@ -1595,6 +1603,66 @@ def _generate_noise_temporal_drift(trs,
         phase = (timepoints / (trs - 1) * cycles * 2 * np.pi) + phaseshift
         noise_drift = np.sin(phase)
 
+    elif basis == 'cos_power_drop':
+
+        # Make a vector counting each TR
+        timepoints = np.linspace(0, trs - 1, trs) * tr_duration
+
+        # Specify the other timing information
+        duration = trs * tr_duration
+
+        # How bases do you have?
+        basis_funcs = int(np.floor(2 * duration))
+
+        noise_drift = np.zeros((timepoints.shape[0], basis_funcs))
+        for basis_counter in list(range(1, basis_funcs + 1)):
+            # What steps do you want to take for this basis function
+            random_phase = np.random.rand() * np.pi * 2
+
+            timepoint_phase = (timepoints / duration * np.pi * basis_counter)
+
+            # In radians, what is the value for each time point
+            timepoints_basis = timepoint_phase + random_phase
+
+            # Store the drift from this basis func
+            noise_drift[:, basis_counter - 1] = np.cos(timepoints_basis)
+
+        def power_drop(r, L, F):
+            # Function to return the drop rate for the power of basis functions
+            # In other words, how much should the weight of each basis function
+            # reduce in order to make the power you retain be above a 0.99
+            # r is the power reduction rate which should be between 0 and 1
+            # L is the duration of the run in seconds
+            # F is period of the cycle in seconds
+
+            percent_retained = 0.99  # What is the percentage of drift retained
+            numerator = 1 - r ** (4 * L / F)  # Power of this period
+            denominator = 1 - r ** (4 * L)  # Power of all periods
```

Comment:
According to my comment on Line 1658, I now guess `r` is the reduction rate of the weight of each frequency component from its previous one. So the power it would be raised to would be a unit-less number, not a duration. But `L` is the duration of the scan, which seems to be should not be a power. I am also not clear why 4 is multiplied. If the amplitude is decreased by `r`, then we can say the power is decreased by `r^2`. So I can understand if 2 is used, but not 4.

Reply:
Hmm, let me make some guess to see if I am understanding correctly. In the denominator, `4 * L` is `2 * (2 * L)`, where `2 * L` is `2 * duration`. This would make sense given the codes above where you have set the highest `basis_counter`  to be `2 * duration`. And the  `4 * L / F` is essentially `2 * (1/F / (1/2L))`, where `1/F` is the frequency corresponding to the period you consider retaining 99% of the power below it, and `1/2L` is the lowest frequency in your basis set. So now both of these make sense to me. My only concern is whether setting the highest frequency as `2*L` makes sense. They are of different units to start with: `2L` is a quantity of time, choosing its number as a frequency seems strange to me.

Reply:
Glad it makes sense, will wait on the above comment to see if this still seems workable